test/jsn_tests.erl

```diff
@@ -669,6 +674,70 @@ from_as_proplist_test_() ->
         [fun from_as_proplist/0 || _ <- lists:seq(1, 5)]}.
 
 
+from_as_proplist_error_test_() ->
+    [?_assertError(badarg, jsn:as_proplist(not_json_term)),
+     ?_assertError(badarg, jsn:as_proplist([{<<"key">>, not_json_term}])),
+     ?_assertError(badarg, jsn:as_proplist([{{error, not_json_key}, <<"value">>}])),
+     ?_assertError(badarg, jsn:as_proplist([{<<"key">>, <<"value">>},
+                                            {key2, [{{error, not_json_key}, <<"value">>}]}])),
+     ?_assertError(badarg, jsn:as_proplist(#{{error, not_json_key} => <<"value">>})),
+     ?_assertError(badarg, jsn:as_proplist(#{<<"key">> => <<"value">>,
+                                             key2 => #{{error, not_json_key} => <<"value">>}})),
+     ?_assertError(badarg, jsn:from_proplist(not_json_term)),
+     ?_assertError(badarg, jsn:from_proplist([{<<"key">>, not_json_term}])),
+     ?_assertError(badarg, jsn:from_proplist([{{error, not_json_key}, <<"value">>}])),
+     ?_assertError(badarg, jsn:from_proplist([{<<"key">>, <<"value">>},
+                                              {key2, [{{error, not_json_key}, <<"value">>}]}])),
+     ?_assertError(badarg, jsn:from_proplist([{<<"key">>, <<"value">>},
+                                              {key2, [{{error, not_json_key}, <<"value">>}]}],
+                                             [{format, struct}]))].
+
+
+from_as_map() ->
+    Map = generate_json_object(7, [{format, map}]),
+    Proplist = jsn:as_proplist(Map),
+    Eep18 = jsn:from_proplist(Proplist, [{format, eep18}]),
+    Struct = jsn:from_proplist(Proplist, [{format, struct}]),
+    ?assertEqual(Map, jsn:from_map(Map)),
+    ?assertEqual(Map, jsn:from_map(Map, [{format, map}])),
+    ?assertEqual(Proplist, jsn:from_map(Map, [{format, proplist}])),
+    ?assertEqual(Eep18, jsn:from_map(Map, [{format, eep18}])),
+    ?assertEqual(Struct, jsn:from_map(Map, [{format, struct}])),
+    ?assertEqual(Map, jsn:as_map(Map)),
+    ?assertEqual(Map, jsn:as_map(Proplist)),
+    ?assertEqual(Map, jsn:as_map(Eep18)),
+    ?assertEqual(Map, jsn:as_map(Struct)),
+    ?assertEqual([Map, Map, Map, Map],
+                 jsn:as_map([Map, Proplist, Eep18, Struct])).
+
+
+%% run the above test several times
+from_as_map_test_() ->
```

Comment:
Really should start working on some tests using proper for this library but for now this is fine